finality-aleph/src/runtime_api.rs

```diff
@@ -63,36 +64,74 @@ where
         }
     }
 
-    fn read_storage<D: Decode>(
+    fn access_storage<D: Decode>(
+        &self,
+        storage_key: Vec<u8>,
+        at_block: BlockHash,
+    ) -> Result<D, ApiError> {
+        let encoded = self
+            .client
+            .storage(at_block, &sc_client_api::StorageKey(storage_key))
+            .map_err(|_| ApiError::StorageAccessFailure)?
+            .ok_or(ApiError::NoStorage)?;
+        D::decode_all(&mut encoded.0.as_ref()).map_err(ApiError::DecodeError)
+    }
+
+    fn read_storage_value<D: Decode>(
         &self,
         pallet: &str,
         item: &str,
         at_block: BlockHash,
     ) -> Result<D, ApiError> {
         let storage_key = [twox_128(pallet.as_bytes()), twox_128(item.as_bytes())].concat();
+        match self.access_storage(storage_key, at_block) {
+            Err(ApiError::NoStorage) => Err(ApiError::NoStorageValue(pallet.into(), item.into())),
+            other => other,
+        }
+    }
 
-        let encoded = match self
-            .client
-            .storage(at_block, &sc_client_api::StorageKey(storage_key))
-        {
-            Ok(Some(e)) => e,
-            _ => return Err(ApiError::NoStorage(pallet.to_string(), item.to_string())),
-        };
-
-        D::decode_all(&mut encoded.0.as_ref()).map_err(ApiError::DecodeError)
+    #[allow(dead_code)]
+    fn read_storage_map<H: StorageHasher, D: Decode, E: Encode>(
+        &self,
+        pallet: &str,
+        item: &str,
+        key: E,
+        at_block: BlockHash,
+    ) -> Result<D, ApiError> {
+        let mut storage_key = [twox_128(pallet.as_bytes()), twox_128(item.as_bytes())].concat();
+        let p = key.using_encoded(H::hash);
```

Comment:
Does `using_encoded` take into account that this should be a hash storage keys with concat(twox64(key), key)? From the test I see it's not, so can one use here something like `Blake2_128Concat` as a `H` when calling this function to mimic real-world example? I know this works as it is, but in reall chain `Blake2_128Concat` is used not just `Twox128` - maybe latter is used as well, I did not check all metadata

Reply:
Not sure if I understand the question. Anything that implements `StorageHasher` can be used here as `H` which includes `Blake2_128Concat`. I'll have `Twox128` changed in tests to something which resembles more the real scenario.

Reply:
Okay, that's what I thought for. Thanks